### DISCUSSION TITLE
Move base_uri from Parser/Serializer config to document prop

### DISCUSSION THREAD

**Discussion by damooo** [pchampin/sophia_rs#155] [Q&A] [answered]
Current parsers and serializers are configured with base_uri as their property.
And thus we have to instantiate those for every doc with different uri.

Instead it would be natural to take base_uri as an attribute of document to be parsed/serialized. Thus one parser/serializer instance can operate over any number of docs. Essentially uri is a property of the document.

For example, json-ld interface defines a `RemoteDocument` abstraction, as a product of uri, content, content-type of a document. This seems a good abstraction.

**Comment by pchampin** [comment, marked as answer]
[`BufRead`](https://doc.rust-lang.org/std/io/trait.BufRead.html) is very pervasive in Rust, so I would rather keep them as a possible input for parsers. And therefore (since `BufRead`s don't carry their origin URL), having a default base IRI in the parser configuration makes sense.

But that being said, having an abstraction similar to JSON-LD's `RemoteDocument`, that attaches metadata (such as origin URL) to the data to be parsed, sounds indeed like a good idea. Parsers implementations could accept either `BufRead` of this abstraction.
